lib/types.ts

```diff
@@ -424,7 +424,13 @@ export type Opaque<Type, Token extends string> = Token extends StringLiteral<Tok
   : never;
 
 /** Easily extract the type of a given object's values */
-export type ValueOf<T> = T[keyof T];
+export type ValueOf<T> = T extends Primitive
+  ? T
+  : T extends any[]
```

Comment:
You check only writable array and tuples, but forgot about readonly arrays:

1. Please add the test for readonly array, e.g. `Assert<IsExact<ValueOf<readonly number[]>, number | undefined>>,`

2. Find a way to distinguish arrays, readonly arrays and tuples

3. Add the check here so tests are passed successfully

Let me know if you need any help for building conditional types

Reply:
> You check only writable array and tuples, but forgot about readonly arrays:

Oh yes totally missed that part, will change the conditional type to `AnyArray` as it will handle both mutable and readonly arrays.

> Find a way to distinguish arrays, readonly arrays, and tuples

I think in the conditional type `AnyArray` covers the case for `Tuple` as well because `ValueOf` should yield `T[number]` for all of them.
Have also added the test cases to distinguish b/w Arrays, readonly arrays, and tuples

```ts
export type ValueOf<T> = T extends Primitive
  ? T
  : T extends AnyArray
  ? T[number]
  : T extends AnyFunction
  ? ReturnType<T>
  : T[keyof T];
````

Reply:
Thank you for adjusting the logic, LGTM!

And yes, we don't need `IsTuple` as the result for arrays, readonly arrays and tuples will be identical!

Reply:
Awesome!